### PR TITLE
Multi script group witness signing

### PR DESCRIPTION
**Pull request by classicalliu**

### CODE REVIEW

File: packages/neuron-wallet/src/services/wallets.ts [hunk 1]
```diff
@@ -371,18 +371,55 @@ export default class WalletService {
...
+        witness: '',
+        blake160,
+      }
+    })
+
+    const lockHashes = new Set(witnessesWithLockHashes.map(w => w.lockHash))
+
+    for (const lockHash of lockHashes) {
+      const firstIndex = witnessesWithLockHashes.findIndex(w => w.lockHash === lockHash)
+      const witnessesArgsWithBlake160 = witnessesWithLockHashes
```
Comment (Keith-CY) [inline packages/neuron-wallet/src/services/wallets.ts]: Now there are `witness`, `witnesses`, `witnessArgs`, `witnessesArgs`, which is bewildering.

File: packages/neuron-wallet/src/services/wallets.ts [hunk 2]
```diff
@@ -372,38 +372,54 @@ export default class WalletService {
...
+        .filter(w => w.lockHash === lockHash)
+        .map(w => ({args: w.witnessArgs, blake160: w.blake160}))
       // A 65-byte empty signature used as placeholder
-      witnesses[0].lock = "0x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
-      const signedWitness = core.signWitness(privateKey)({
+      witnessesArgsWithBlake160[0].args.lock = '0x' + '0'.repeat(130)
+
+      const blake160: string = witnessesArgsWithBlake160[0].blake160
```
Comment (ashchan) [inline packages/neuron-wallet/src/services/wallets.ts]: I feel we should extract this to a function to isolate the private key fetch.

File: packages/neuron-wallet/src/services/wallets.ts [hunk 3]
```diff
@@ -371,18 +371,55 @@ export default class WalletService {
     const paths = addressInfos.map(info => info.path)
     const pathAndPrivateKeys = this.getPrivateKeys(wallet, paths, password)
 
-    const witnesses: string[] = inputs!.map((input: Input) => {
+    const witnessesWithLockHashes = inputs!.map((input: Input) => {
```
Comment (ashchan) [inline packages/neuron-wallet/src/services/wallets.ts]: `witnessesWithLockHashes` is really a bad name. If I understand how it was named previously, it should be called `witnessesWithLockHashesAndWitnessArgsAndBlake160` now?

File: packages/neuron-wallet/src/services/wallets.ts (hunk 1 above)
Comment (ashchan) [reply packages/neuron-wallet/src/services/wallets.ts]: `witnessesWithLockHashes` becomes a bad name now that there're more members.. If I understand how it was named previously, it should be called `witnessesWithLockHashesAndWitnessArgsAndBlake160` now? :P

Comment (xxuejie) [reply packages/neuron-wallet/src/services/wallets.ts]: Or just witnessGenerationData for short

**Pull request merged**